File: src/main/java/com/naminhyeok/fantazzk/template/application/ProvideTemplateCatalog.java
package com.naminhyeok.fantazzk.template.application;

import com.naminhyeok.fantazzk.CoreException;
import com.naminhyeok.fantazzk.template.TemplateCatalog;
import com.naminhyeok.fantazzk.template.domain.Template;
import com.naminhyeok.fantazzk.template.domain.TemplateErrorType;
import com.naminhyeok.fantazzk.template.domain.TemplateId;
import com.naminhyeok.fantazzk.template.query.FindTemplates;
import com.naminhyeok.fantazzk.template.query.TemplateDetail;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class ProvideTemplateCatalog implements TemplateCatalog {
    private final FindTemplates findTemplates;

    @Override
    public TemplateBlueprint getTemplate(UUID templateId) {
        try {
            TemplateDetail detail = findTemplates.getDetail(new TemplateId(templateId));
            Template template = detail.template();
            return new TemplateBlueprint(
                template.getGameType(),
                template.getMode(),
                template.getTeamCount(),
                template.getTeamSize(),
                template.getBudget(),
                template.getPickBanTime(),
                template.getMinBidUnit(),
                template.getDraftOrderStrategy(),
                detail.players().stream()
                    .map(player -> new PlayerBlueprint(player.name(), player.position(), player.displayOrder()))
                    .toList()
            );
        } catch (CoreException ex) {
            if (ex.getError() == TemplateErrorType.TEMPLATE_NOT_FOUND) {
                throw new TemplateCatalog.NotFound(templateId);
            }
            throw ex;
        }
    }
}

File: src/main/kotlin/com/naminhyeok/fantazzk/template/application/ProvideTemplateCatalog.kt
package com.naminhyeok.fantazzk.template.application

import com.naminhyeok.fantazzk.template.TemplateBlueprint
import com.naminhyeok.fantazzk.template.TemplateCatalog
import com.naminhyeok.fantazzk.template.TemplateCatalogException
import com.naminhyeok.fantazzk.template.TemplateDraftOrderStrategy
import com.naminhyeok.fantazzk.template.TemplateId
import com.naminhyeok.fantazzk.template.TemplateMode
import com.naminhyeok.fantazzk.template.TemplatePlayerBlueprint
import com.naminhyeok.fantazzk.template.exception.TemplateException
import org.springframework.stereotype.Component

@Component
internal class ProvideTemplateCatalog(
    private val templateFinder: FindTemplates,
) : TemplateCatalog {
    override fun getTemplateBlueprint(templateId: TemplateId): TemplateBlueprint =
        try {
            val detail = templateFinder.getDetail(templateId)
            TemplateBlueprint(
                templateId,
                TemplateMode.valueOf(detail.template.mode.name),
                detail.template.teamCount,
                detail.template.teamSize,
                detail.template.budget,
                detail.template.draftOrderStrategy?.let {
                    TemplateDraftOrderStrategy.valueOf(it.name)
                },
                detail.players.map {
                    TemplatePlayerBlueprint(
                        it.name,
                        it.displayOrder,
                    )
                },
            )
        } catch (_: TemplateException.TemplateNotFoundException) {
            throw TemplateCatalogException.NotFound(templateId)
        } catch (_: TemplateException.TemplateInvalidException) {
            throw TemplateCatalogException.Invalid(templateId)
        }
}
